### PR TITLE
Support for cross zone mounting

### PR DESCRIPTION
**Pull request by sameshai**

### CODE REVIEW

**Comment by sameshai** [comment]
```
sameershaikh@Sameers-MacBook-Pro eni % kubectl apply -f custom-pvc-eni.yaml        
persistentvolumeclaim/pv-cross created
sameershaikh@Sameers-MacBook-Pro eni % kubectl apply -f pvc-secret.yaml            
persistentvolumeclaim/volume-secret created
sameershaikh@Sameers-MacBook-Pro eni % kubectl get pvc                                   
NAME            STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS                  AGE
pv-cross        Pending                                      ibmc-vpc-file-dp2-cross       23s
volume-secret   Pending                                      ibmc-vpc-file-dp-eni-custom   15s
sameershaikh@Sameers-MacBook-Pro eni % kubectl get pvc
NAME            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                  AGE
pv-cross        Bound    pvc-49453803-abf9-4d98-839a-784aa9c58249   10Gi       RWX            ibmc-vpc-file-dp2-cross       54s
volume-secret   Bound    pvc-af61f8bf-d7cd-4287-bee9-5009afae1847   10Gi       RWX            ibmc-vpc-file-dp-eni-custom   46s
sameershaikh@Sameers-MacBook-Pro eni % kubectl describe pv                               
Name:              pvc-49453803-abf9-4d98-839a-784aa9c58249
Labels:            <none>
Annotations:       pv.kubernetes.io/provisioned-by: vpc.file.csi.ibm.io
                   volume.kubernetes.io/provisioner-deletion-secret-name: 
                   volume.kubernetes.io/provisioner-deletion-secret-namespace: 
Finalizers:        [kubernetes.io/pv-protection]
StorageClass:      ibmc-vpc-file-dp2-cross
Status:            Bound
Claim:             default/pv-cross
Reclaim Policy:    Delete
Access Modes:      RWX
VolumeMode:        Filesystem
Capacity:          10Gi
Node Affinity:     
  Required Terms:  
    Term 0:        failure-domain.beta.kubernetes.io/region in [us-south]
Message:           
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            vpc.file.csi.ibm.io
    FSType:            
    VolumeHandle:      r134-ad4328fb-fe00-45a9-b32a-112d3a9ab46e:r134-ba474b0d-bcbb-4b82-8590-efd98c90750f
    ReadOnly:          false
    VolumeAttributes:      clusterID=cj7mror20g32qcvcud00
                           failure-domain.beta.kubernetes.io/region=us-south
                           iops=100
                           nfsServerPath=10.240.0.30:/e39711a4_6ba8_45cb_aa04_415c622b76b2
                           storage.kubernetes.io/csiProvisionerIdentity=1697742530710-8081-vpc.file.csi.ibm.io
                           tags=
                           volumeCRN=crn:v1:staging:public:is:us-south-1:a/77f2bceddaeb577dcaddb4073fe82c1c::share:r134-ad4328fb-fe00-45a9-b32a-112d3a9ab46e
                           volumeId=r134-ad4328fb-fe00-45a9-b32a-112d3a9ab46e:r134-ba474b0d-bcbb-4b82-8590-efd98c90750f
Events:
  Type     Reason               Age   From                                        Message
  ----     ------               ----  ----                                        -------
  Warning  VolumeMetaDataSaved  4s    ibm-vpc-file-csi-controller-5fb54554-ztx8z  unsupported Operation
  Warning  VolumeMetaDataSaved  4s    ibm-vpc-file-csi-controller-5fb54554-rl6jn  unsupported Operation


Name:              pvc-af61f8bf-d7cd-4287-bee9-5009afae1847
Labels:            <none>
Annotations:       pv.kubernetes.io/provisioned-by: vpc.file.csi.ibm.io
                   volume.kubernetes.io/provisioner-deletion-secret-name: volume-secret
                   volume.kubernetes.io/provisioner-deletion-secret-namespace: default
Finalizers:        [kubernetes.io/pv-protection]
StorageClass:      ibmc-vpc-file-dp-eni-custom
Status:            Bound
Claim:             default/volume-secret
Reclaim Policy:    Delete
Access Modes:      RWX
VolumeMode:        Filesystem
Capacity:          10Gi
Node Affinity:     
  Required Terms:  
    Term 0:        failure-domain.beta.kubernetes.io/region in [us-south]
                   failure-domain.beta.kubernetes.io/zone in [us-south-1]
Message:           
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            vpc.file.csi.ibm.io
    FSType:            
    VolumeHandle:      r134-261da83f-9b5b-476e-862f-4e32841111cb:r134-aeaf7104-8df2-4b54-a14c-e11d2419ca0f
    ReadOnly:          false
    VolumeAttributes:      clusterID=cj7mror20g32qcvcud00
                           failure-domain.beta.kubernetes.io/region=us-south
                           failure-domain.beta.kubernetes.io/zone=us-south-1
                           iops=500
                           nfsServerPath=fsf-dal1095b-vs91-stg-fz.adn.networklayer.com:/c344d198_c406_4e4b_a80c_ef10a4fbc0ed
                           storage.kubernetes.io/csiProvisionerIdentity=1697742530710-8081-vpc.file.csi.ibm.io
                           tags=
                           volumeCRN=crn:v1:staging:public:is:us-south-1:a/77f2bceddaeb577dcaddb4073fe82c1c::share:r134-261da83f-9b5b-476e-862f-4e32841111cb
                           volumeId=r134-261da83f-9b5b-476e-862f-4e32841111cb:r134-aeaf7104-8df2-4b54-a14c-e11d2419ca0f
Events:
  Type     Reason               Age   From                                        Message
  ----     ------               ----  ----                                        -------
  Warning  VolumeMetaDataSaved  13s   ibm-vpc-file-csi-controller-5fb54554-rl6jn  unsupported Operation
  Warning  VolumeMetaDataSaved  12s   ibm-vpc-file-csi-controller-5fb54554-ztx8z  unsupported Operation
sameershaikh@Sameers-MacBook-Pro eni %                                             
sameershaikh@Sameers-MacBook-Pro eni %                                     
sameershaikh@Sameers-MacBook-Pro eni %  kubectl delete -f custom-deployment.yaml 
^C
sameershaikh@Sameers-MacBook-Pro eni %  kubectl apply -f custom-deployment.yaml    
deployment.apps/testpodrwm-dp2 created
sameershaikh@Sameers-MacBook-Pro eni % kubectl get pods -o wide                          
NAME                             READY   STATUS    RESTARTS   AGE   IP              NODE           NOMINATED NODE   READINESS GATES
testpodrwm-dp2-96595f8d7-f964p   1/1     Running   0          5s    172.17.1.108    10.240.64.32   <none>           <none>
testpodrwm-dp2-96595f8d7-m769v   1/1     Running   0          5s    172.17.1.107    10.240.64.32   <none>           <none>
testpodrwm-dp2-96595f8d7-mfpkv   1/1     Running   0          5s    172.17.31.210   10.240.0.63    <none>           <none>
testpodrwm-dp2-96595f8d7-sbds4   1/1     Running   0          5s    172.17.31.209   10.240.0.63    <none>           <none>
testpodrwm-dp2-96595f8d7-z9szv   1/1     Running   0          5s    172.17.31.202   10.240.0.63    <none>           <none>
sameershaikh@Sameers-MacBook-Pro eni %                                           
sameershaikh@Sameers-MacBook-Pro eni % cp custom-deployment.yaml custom-deployment-vpc.yaml
sameershaikh@Sameers-MacBook-Pro eni % vi custom-deployment-vpc.yaml 
sameershaikh@Sameers-MacBook-Pro eni %  kubectl apply -f custom-deployment.yaml 
sameershaikh@Sameers-MacBook-Pro eni % 
sameershaikh@Sameers-MacBook-Pro eni % vi custom-deployment-vpc.yaml                       
sameershaikh@Sameers-MacBook-Pro eni %  kubectl apply -f custom-deployment-vpc.yaml 
deployment.apps/testpodrwm-dp2-vpc created
sameershaikh@Sameers-MacBook-Pro eni % kubectl get pods -o wide
NAME                                  READY   STATUS    RESTARTS   AGE     IP              NODE           NOMINATED NODE   READINESS GATES
testpodrwm-dp2-96595f8d7-f964p        1/1     Running   0          2m10s   172.17.1.108    10.240.64.32   <none>           <none>
testpodrwm-dp2-96595f8d7-m769v        1/1     Running   0          2m10s   172.17.1.107    10.240.64.32   <none>           <none>
testpodrwm-dp2-96595f8d7-mfpkv        1/1     Running   0          2m10s   172.17.31.210   10.240.0.63    <none>           <none>
testpodrwm-dp2-96595f8d7-sbds4        1/1     Running   0          2m10s   172.17.31.209   10.240.0.63    <none>           <none>
testpodrwm-dp2-96595f8d7-z9szv        1/1     Running   0          2m10s   172.17.31.202   10.240.0.63    <none>           <none>
testpodrwm-dp2-vpc-7748f75955-dp47b   1/1     Running   0          6s      172.17.31.212   10.240.0.63    <none>           <none>
testpodrwm-dp2-vpc-7748f75955-td9l7   1/1     Running   0          6s      172.17.31.215   10.240.0.63    <none>           <none>
testpodrwm-dp2-vpc-7748f75955-vjhqt   1/1     Running   0          6s      172.17.31.216   10.240.0.63    <none>           <none>
testpodrwm-dp2-vpc-7748f75955-w5jp7   1/1     Running   0          6s      172.17.31.214   10.240.0.63    <none>           <none>
testpodrwm-dp2-vpc-7748f75955-wtp8g   1/1     Running   0          6s      172.17.31.211   10.240.0.63    <none>           <none>
sameershaikh@Sameers-MacBook-Pro eni % kubectl get nodes --show-labels
NAME           STATUS   ROLES    AGE   VERSION       LABELS
10.240.0.63    Ready    <none>   35d   v1.27.4+IKS   arch=amd64,beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=cx2.2x4,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-south,failure-domain.beta.kubernetes.io/zone=us-south-1,ibm-cloud.kubernetes.io/ha-worker=true,ibm-cloud.kubernetes.io/iaas-provider=g2,ibm-cloud.kubernetes.io/instance-id=0716_c8776eff-9c13-4cd2-908f-8720191bf94d,ibm-cloud.kubernetes.io/internal-ip=10.240.0.63,ibm-cloud.kubernetes.io/machine-type=cx2.2x4,ibm-cloud.kubernetes.io/os=UBUNTU_20_64,ibm-cloud.kubernetes.io/region=us-south,ibm-cloud.kubernetes.io/sgx-enabled=false,ibm-cloud.kubernetes.io/subnet-id=0716-8b4060f9-2464-4710-9c87-bc4b16a1d72f,ibm-cloud.kubernetes.io/worker-id=test-cj7mror20g32qcvcud00-sushma127-default-00002dce,ibm-cloud.kubernetes.io/worker-pool-id=cj7mror20g32qcvcud00-27e3804,ibm-cloud.kubernetes.io/worker-pool-name=default,ibm-cloud.kubernetes.io/worker-version=1.27.4_1538,ibm-cloud.kubernetes.io/zone=us-south-1,kubernetes.io/arch=amd64,kubernetes.io/hostname=10.240.0.63,kubernetes.io/os=linux,node.kubernetes.io/instance-type=cx2.2x4,topology.kubernetes.io/region=us-south,topology.kubernetes.io/zone=us-south-1
10.240.64.32   Ready    <none>   16d   v1.27.5+IKS   arch=amd64,beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=cx2.2x4,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-south,failure-domain.beta.kubernetes.io/zone=us-south-2,ibm-cloud.kubernetes.io/ha-worker=true,ibm-cloud.kubernetes.io/iaas-provider=g2,ibm-cloud.kubernetes.io/instance-id=0726_5b9a1db6-a751-4779-942d-9f4ffbbf14a2,ibm-cloud.kubernetes.io/internal-ip=10.240.64.32,ibm-cloud.kubernetes.io/machine-type=cx2.2x4,ibm-cloud.kubernetes.io/os=UBUNTU_20_64,ibm-cloud.kubernetes.io/region=us-south,ibm-cloud.kubernetes.io/sgx-enabled=false,ibm-cloud.kubernetes.io/subnet-id=0726-6fd4b505-793f-47d3-8879-b9b52dd6c71c,ibm-cloud.kubernetes.io/worker-id=test-cj7mror20g32qcvcud00-sushma127-default-00002e61,ibm-cloud.kubernetes.io/worker-pool-id=cj7mror20g32qcvcud00-27e3804,ibm-cloud.kubernetes.io/worker-pool-name=default,ibm-cloud.kubernetes.io/worker-version=1.27.5_1541,ibm-cloud.kubernetes.io/zone=us-south-2,kubernetes.io/arch=amd64,kubernetes.io/hostname=10.240.64.32,kubernetes.io/os=linux,node.kubernetes.io/instance-type=cx2.2x4,topology.kubernetes.io/region=us-south,topology.kubernetes.io/zone=us-south-2
sameershaikh@Sameers-MacBook-Pro eni % 

```

File: pkg/ibmcsidriver/controller_helper.go
Comment (arahamad) [inline pkg/ibmcsidriver/controller_helper.go:307]: can this be optimized, in case of subnetID or primaryIPID is not given then also we are checking zone & region which is not required

**Review by arahamad — APPROVED** [review]
lgtm, lets verify after merging all PRs

File: pkg/ibmcsidriver/controller_helper.go
Comment (sameshai) [reply pkg/ibmcsidriver/controller_helper.go:307]: We will take in next PR and make it case wise

**Review by prankulmahajan — APPROVED** [review]
lgtm

**Pull request merged**
